### PR TITLE
Mixed precision (MP)

### PR DESCRIPTION
**Pull request by markus-hinsche**
**Motivation**: Mixed precision is the use of both 16-bit and 32-bit floating-point types in a model during training to make it run faster and use less memory.

Works only for some GPU types, e.g. T4 or V100 (https://www.kaggle.com/discussion/198232) (https://www.tensorflow.org/guide/mixed_precision#supported_hardware mentions that there is a speedup for V100. )

TODO compare GPU memory

**Resources:**
- https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html
- https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html#scalefactor
- https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/LossScaleOptimizer
- https://www.tensorflow.org/guide/mixed_precision
- https://developer.nvidia.com/automatic-mixed-precision
- The code is taken from here: https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/chapter13_best-practices-for-the-real-world.ipynb
- Note that some operations may be numerically unstable in float16 (in particular, softmax and crossentropy). If you need to opt out of mixed precision for a specific layer, just pass the argument dtype="float32" to the constructor of this layer.

# MP speeds up training on V100

| Run       | Comment        | Epoch duration  |
| -------- |:-------------:| -----:|
| [Run1](https://ml.azure.com/experiments/id/0d5557e4-e7ff-4c55-b53d-e7d0edeea945?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourceGroups/cgm-ml-prod-ci-rg/providers/Microsoft.MachineLearningServices/workspaces/cgm-ml-prod-ci-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b)      | mixed precision                   | `61s` |
| [Run12](https://ml.azure.com/experiments/id/0d5557e4-e7ff-4c55-b53d-e7d0edeea945?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourceGroups/cgm-ml-prod-ci-rg/providers/Microsoft.MachineLearningServices/workspaces/cgm-ml-prod-ci-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b)      | baseline (F32)                   | `96s` |

**Results:**
- 1.6 times as fast
- BUT loss curve is much worse

# MP doesn't speed up training on K80
[Experiment](https://ml.azure.com/experiments/id/f3e2b900-5d3e-4900-9e10-841a8766cd49?wsid=/subscriptions/9b5bbfae-d5d1-4aae-a2ca-75159c0c887d/resourcegroups/cgm-ml-prod-we-rg/workspaces/cgm-ml-prod-we-azml&tid=3a27c573-ec1a-4734-9cd3-3208af51794b):

| Run       | Comment        | Epoch duration  |
| -------- |:-------------:| -----:|
| Run11      | mixed precision                   | 678s |
| Run12      | baseline (F32)                   | 656s |

**Result:** 
- not faster on K80 with TF2.4 -> I would be nice to try on a V100
- also MAE got very high for the mixed precision run

**Potential future work:** 
- check dtype (`float32`, `float16`)

### CODE REVIEW

**Comment by markus-hinsche** [comment]
Doesn't work for our GPUs, could be revisited once we can use other GPUs
